### PR TITLE
Minor: Improve `PruningPredicate` documentation

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A

## Rationale for this change

We make use of the PruningPredicate a lot in IOx and I think it is a powerful feature. However, it is hard to explain to people / refer them to docs about it. 

I would like to make it easier to understand by reading the docs

## What changes are included in this PR?

Improve the docs for PruningPredicate, clarifying the rationale as well as adding some examples

## Are these changes tested?

By doc tests

## Are there any user-facing changes?
Better docs. No code changes

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 1]
```diff
@@ -146,17 +172,14 @@ impl PruningPredicate {
     ///
     /// `true`: There MAY be rows that match the predicate
     ///
-    /// `false`: There are no rows that could match the predicate
+    /// `false`: There are no rows that could possible match the predicate
```
Comment (mhilton) [inline datafusion/core/src/physical_optimizer/pruning.rs:175]: *possibly

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 2]
```diff
@@ -91,10 +83,44 @@ pub trait PruningStatistics {
...
+/// [`PruningPredicate`]s are used to prune (avoid scanning) Parquet Row Groups
+/// based on the min/max values found in the Parquet metadata. If the
+/// `PruningPredicate` can guarantee that no rows in the Row Group match the
+/// filter, the entire Row Group is skipped during query execution.
+///
+/// Note that this API is designed to be general, as it works:
+///
+/// 1. Arbitrary expressions expressions (including user defined functions)
+///
+/// 2. Anything that implements the [`PruningStatistics`] interface, not just
```
Comment (mhilton) [inline datafusion/core/src/physical_optimizer/pruning.rs:100]: Should "interface" be "trait" in this context?

**Review by mhilton — APPROVED** [review]

File: datafusion/core/src/physical_optimizer/pruning.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/physical_optimizer/pruning.rs:100]: Yes, you are correct. Thank you

Comment (alamb) [reply datafusion/core/src/physical_optimizer/pruning.rs:100]:
```suggestion
/// 2. Anything that implements the [`PruningStatistics`] trait, not just
```

**Review by jackwener — APPROVED** [review]

**Comment by alamb** [comment]
Thank you @jackwener

**Pull request merged**